ocaml/xapi/storage_access.ml

```diff
@@ -454,9 +454,11 @@ module SMAPIv1 = struct
                      extra = attach_info_v1.Smint.xenstore_data;
                      backend_type = "vbd3"
                    };
-                   (* Currently we always get a BlockDevice from SMAPIv1, never a File, not even for ISOs *)
                    BlockDevice {
```

Comment:
You can just drop BlockDevice from here, and use Nbd.
Then you don't need to change the order of priority in xenopsd.

Reply:
Although @MarkSymsCtx  just mentioned CDROMs where we do need a blockdevice.

Reply:
Ideally the SM backend would also specify priorities instead of having XAPI guess.

If everything worked correctly then the most efficient ways to access the device would be:
* Path (O_DIRECT)
* Blockdevice (O_DIRECT)
* NBD

However we currently prefer NBD over Blockdevice because tapdisk's blockdevice support is something we want to remove, and something that uses pagecache when it shouldn't.
Blockdevice is also currently not more efficient than NBD because it apparently wouldn't use O_DIRECT. However that sounds like something to fix in the SM layer.

Thus currently NBD is preferrable to BlockDevices when used for Qemu. However when used for import/export in Dom0 (or any other reason why it might be plugged into Dom0) then NBD would spawn an nbd-client, and we'd get a blockdevice  (which would run an extra process, and allocate extra memory for the buffer, although hopefully a fixed size and not unbounded page-cache).

Seems like XAPI/xenopsd would gain a lot of implementation-specific workarounds (for CD-ROMs, blockdevice memory usage, etc.), where really the implementation would know what is most efficient and tell XAPI about that.

For SMAPIv3 we could change the API to say that the order you return the implementations in is the preferred priority order (except XenDisk would only be used for VMs).
For SMAPIv1 unfortunately the API doesn't specify a list of implementations, there are just various parameters.

Since we're adding a new optional parameter to SMAPIv1, could we instead make it return the SMAPIv3 implementation list then?
i.e. instead of 
```ocaml
params: string
param_nbd: string option
```

Have this:
```
implementations: Storage_interface.implementation list
```

And in parse_attach_result make `params` optional and look for `implementations` in the XML, and if found transform that to an `implementation list`. This would be more work, but would make SMAPIv1 closer to SMAPIv3 and add more flexibility for the backend to make appropriate choices. 
E.g. an ISOSR could always return a blockdevice (or even a file) to work around issues with Qemu and fd passing, and once that is fixed it could revert to returning NBD again.